src/Core/Grid/Definition/Factory/CmsPageCategoryDefinitionFactory.php

```diff
@@ -0,0 +1,318 @@
+use PrestaShop\PrestaShop\Core\Grid\Action\GridActionCollection;
+use PrestaShop\PrestaShop\Core\Grid\Action\Row\RowActionCollection;
+use PrestaShop\PrestaShop\Core\Grid\Action\Row\Type\LinkRowAction;
+use PrestaShop\PrestaShop\Core\Grid\Action\Row\Type\SubmitRowAction;
+use PrestaShop\PrestaShop\Core\Grid\Action\Type\SimpleGridAction;
+use PrestaShop\PrestaShop\Core\Grid\Column\ColumnCollection;
+use PrestaShop\PrestaShop\Core\Grid\Column\Type\Catalog\PositionColumn;
+use PrestaShop\PrestaShop\Core\Grid\Column\Type\Common\ActionColumn;
+use PrestaShop\PrestaShop\Core\Grid\Column\Type\Common\BulkActionColumn;
+use PrestaShop\PrestaShop\Core\Grid\Column\Type\Common\LinkColumn;
+use PrestaShop\PrestaShop\Core\Grid\Column\Type\Common\ToggleColumn;
+use PrestaShop\PrestaShop\Core\Grid\Column\Type\DataColumn;
+use PrestaShop\PrestaShop\Core\Grid\Filter\Filter;
+use PrestaShop\PrestaShop\Core\Grid\Filter\FilterCollection;
+use PrestaShopBundle\Form\Admin\Type\SearchAndResetType;
+use PrestaShopBundle\Form\Admin\Type\YesAndNoChoiceType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Class CmsPageCategoryDefinitionFactory builds Grid definition for Cms page category listing.
+ */
+final class CmsPageCategoryDefinitionFactory extends AbstractGridDefinitionFactory
+{
+    /**
+     * @var int
+     */
+    private $cmsCategoryParentId;
+
+    /**
+     * @param RequestStack $requestStack
+     */
+    public function __construct(RequestStack $requestStack)
+    {
+        $this->setCmsPageCategoryParentId($requestStack);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getId()
+    {
+        return 'cms_page_category';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getName()
+    {
+        return $this->trans('Categories', [], 'Admin.Navigation.Menu');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getColumns()
+    {
+        return (new ColumnCollection())
+            ->add((new BulkActionColumn('bulk'))
+                ->setOptions([
+                    'bulk_field' => 'id_cms_category',
+                ])
+            )
+            ->add((new DataColumn('id_cms_category'))
+                ->setName($this->trans('ID', [], 'Admin.Global'))
+                ->setOptions([
+                    'field' => 'id_cms_category',
+                ])
+            )
+            ->add((new LinkColumn('name'))
+                ->setName($this->trans('Name', [], 'Admin.Global'))
+                ->setOptions([
+                    'field' => 'name',
+                    'route' => 'admin_cms_pages_edit_cms_category',
+                    'route_param_name' => 'cmsCategoryId',
+                    'route_param_field' => 'id_cms_category',
+                    'route_params_extra' => [
+                        'id_parent' => 'cmsCategoryParentId',
```

Comment:
@matks If I'm using query parameter such as `id_cms_category` when generating lists I want to keep thay query parameter no matter what if it exists. If customer goes to edit page of the cms category then, instead of just having `id_cms_category` I need to be aware also of the parent `id_cms_category` so the client after form submit can return to the parent list and not to the very top list.

The `LinkColumn` was very limited to just passing singe route parameter.
Do you think `route_params_extra` approach is good? Maybe i only need to rename it to `route_params` or something?

Reply:
You mean the following scenario ?
- you are on grid, showing CMS pages & categories of Category "Clothes" which is a subcategory of "Home"
- you click on "edit" for category "Shirts" which is a subcategory of "Clothes"
- you edit this category "Shirts"
- when you submit you go back to grid, showing CMS pages & categories of Category "Clothes"

Is that right ?

I checked the legacy behavior: currently it uses the CMS Category ID of the page being edited. So on last step, instead of showing the "Clothes" grid, the legacy code shows the "Shirts" category.
Changing this behavior can be a good idea but we need Product approval and it deserves its own issue / PR 😄 . Can you create one and ping colinegin for it ?

Reply:
We have an issue with Add/Edit cms page category. I guess when someone will raise the PR which is related with that issue then  the product team can decide what to do. Right now I have one clean solution in mind which does not require to introduce any `route_params_extra` stuff and can be easily changed if product team will desire to change where page should be redirected after the add/edit action  :+1: